### PR TITLE
feat: Charge intrinsic gas on op duration throttle breach

### PR DESCRIPTION
**Pull request by lukelee-sl**
**Description**:
Currently, when the ops duration throttle is reached during transaction processing, the entire gas limit sent in the transaction is charged.  The correct behavior is the charge only the intrinsic gas in this case.  In order to do this, we need to reset the gas charging events (if any) stored while processing the transaction so that the replay of the gas charging that occurs when an exception happen in contract transactions can contain just the single event for the charge of the intrinsic gas.

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by stoyanov-st — DISMISSED** [review]
LG for the solution 
It still needs some unit test coverage and some negative Jumbo tests needs to be adjusted with proper gas expectations.
tyvm @lukelee-sl

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java [hunk 1]
```diff
@@ -116,10 +117,7 @@ public CallOutcome call() {
         final var hevmTransaction = safeCreateHevmTransaction();
         if (hevmTransaction.isException()) {
             return maybeChargeFeesAndReturnOutcome(
-                    hevmTransaction,
-                    context.body().transactionIDOrThrow().accountIDOrThrow(),
-                    null,
-                    contractsConfig.chargeGasOnEvmHandleException());
+                    hevmTransaction, context.payer(), null, contractsConfig.chargeGasOnEvmHandleException());
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java:120]: Q: Are`context.payer()` and `context.body().transactionIDOrThrow().accountIDOrThrow()` always equal?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java [hunk 2]
```diff
@@ -239,4 +243,17 @@ private void setOpsDurationValues() {
...
         hederaOpsDuration.applyDurationFromConfig(requireNonNull(opsDurationConfig));
     }
+
+    /**
+     * Determines whether we should charge gas for exceptions after processing has started.
+     *
+     * @param hevmTransaction the Hedera EVM transaction
+     * @return true if gas should be charged, false otherwise
+     */
+    private boolean shouldChargeGasAfterProcessing(@NonNull final HederaEvmTransaction hevmTransaction) {
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java]: nitpicking: maybe `shouldChargeGasAfterExceptionalProcessing`?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java [hunk 3]
```diff
@@ -195,12 +193,14 @@ private CallOutcome maybeChargeFeesAndReturnOutcome(
...
+
+        long gasCharged = 0;
         if (chargeGas) {
-            gasCharging.chargeGasForAbortedTransaction(
+            gasCharged = gasCharging.possiblyChargeGasForAbortedTransaction(
                     senderId, hederaEvmContext, rootProxyWorldUpdater, hevmTransaction);
         }
 
+        final var status = requireNonNull(hevmTransaction.exception()).getStatus();
         rootProxyWorldUpdater.commit();
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java:204]: Not a big fan of the method naming in this class in general, e.g. `maybeChargeFeesAndReturnOutcome` actually does the commit, which is arguably the most important part of its body, yet the name doesn't indicate that.
Just a note, no need to rename in this PR :)

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCharging.java [hunk 4]
```diff
@@ -153,15 +155,51 @@ public void chargeGasForAbortedTransaction(
...
-        final var intrinsicGas = gasCalculator.transactionIntrinsicGasCost(transaction.evmPayload(), false);
+        /*
+         * There are several possibilities to consider here:
+         * 1. The transaction could have been aborted during parsing before any gas was charged. In this case charge intrinsic gas to payer if possible.
+         * 2. The transaction could have been aborted with a ResourceExhaustedException, which means we need to remove all gas charging events and
+         *     a. If the transaction is an Ethereum transaction, charge the relayer or sender for the intrinsic gas based on gas price.
+         *     b. otherwise, charge the sender for the intrinsic gas.
+         * In all other cases do nothing as the gas charging events are correct - even in the case of a non ResourceExhaustedException exception
+         * in which case the refund is already handled in the {@link FrameRunner#runToCompletion} method.
+         *
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCharging.java:166]: Nice, helpful comment 👍

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCharging.java [hunk 5]
```diff
@@ -153,15 +155,51 @@ public void chargeGasForAbortedTransaction(
...
+        }
+        if (transaction.exception() instanceof ResourceExhaustedException) {
+            worldUpdater.resetGasChargingEvents();
+            if (transaction.isEthereumTransaction()) {
+                final var senderAccount = worldUpdater.getHederaAccount(sender);
+                final var relayerAccount = worldUpdater.getHederaAccount(requireNonNull(transaction.relayerId()));
+                if (senderAccount == null || relayerAccount == null) {
+                    return chargeSenderIntrinsicGas(sender, context, worldUpdater, intrinsicGas);
+                }
+                chargeWithRelayer(senderAccount, relayerAccount, context, worldUpdater, transaction, intrinsicGas);
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCharging.java]: `chargeWithRelayer` seems to be expecting an Hbar (or USD? not sure how gas cost is denominated at this point) value, but we're passing `intrinsicGas` in gas units. Is that correct?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/scope/HandleHederaOperations.java [hunk 6]
```diff
@@ -259,6 +259,29 @@ public void refundGasFee(@NonNull final AccountID payerId, final long amount) {
...
+        gasChargingEvents.clear();
+    }
+
+    @Override
+    public boolean hasGasChargingEvents() {
+        return !gasChargingEvents.isEmpty();
+    }
+
+    @Override
+    public boolean hasRefundGasFeeEvents() {
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/scope/HandleHederaOperations.java:276]: This doesn't seem to be utilized (or I'm missing it) - perhaps some leftover that could be removed?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransaction.java [hunk 7]
```diff
@@ -126,7 +126,7 @@ public HederaEvmTransaction withException(@NonNull final HandleException excepti
                 this.chainId,
                 this.value,
                 this.gasLimit,
-                this.offeredGasPrice,
+                0, // In the case of an exception
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransaction.java:129]: Not sure I understand why we're dropping the original `offeredGasPrice` here?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java (hunk 1 above)
Comment (lukelee-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java:120]: This is in the case where the transaction was not able to be parsed correctly as a smart contracts transaction and so the results here are a bit unpredictable.  `context.payer` will is the results of the payer taking into account the exact transaction type - `ContractCall`, `ContractCreate`, or `EthereumTransaction` and thus I felt it was more accurate the account on the transaction.

Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java:120]: Got it, makes sense. Thanks for clarifying.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCharging.java (hunk 5 above)
Comment (lukelee-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCharging.java]: I was wondering this as well.  I see where the translation is being done and added it to the amount.  The problem is that I still needed to return the gas amount (rather than the fee amount) to the caller so that the record stream result can be updated.

Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCharging.java]: Wouldn't it suffice if we brought back `gasCost = Math.multiplyExact(gasToCharge, context.gasPrice())` in `chargeWithRelayer`?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java [hunk 8]
```diff
@@ -239,4 +243,17 @@ private void setOpsDurationValues() {
...
+     * Determines whether we should charge gas for exceptions after processing has started.
+     *
+     * @param hevmTransaction the Hedera EVM transaction
+     * @return true if gas should be charged, false otherwise
+     */
+    private boolean shouldChargeGasAfterExceptionalProcessing(@NonNull final HederaEvmTransaction hevmTransaction) {
+        return contractsConfig.chargeGasOnEvmHandleException()
+                &&
+                // we charge gas for all contract call and on ResourceExhaustedException
+                (hevmTransaction.isContractCall() || hevmTransaction.exception() instanceof ResourceExhaustedException);
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java:257]: Should we additionally check for a specific `ResourceExhaustedExcaption` status of `THROTTLED_AT_CONSENSUS` or do we want to treat all of them in this manner?
This exception seems to be used for other limits too. I saw a few in `HederaWorldUpdater.commit` which seem to be caught and handled earlier in `TransactionProcessor.safeCommit`, but I'm not sure if some of the other ones can reach it up here.

TBH I find it a bit hard to track which exception is being handled where 😅

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCharging.java (hunk 5 above)
Comment (lukelee-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCharging.java]: I did bring it back while calling `chargeWithRelayer` and inside `chargeSenderIntrinsicGas`.  I did it this way as I still needed to return the `gasUsed` rather than the HAPI fees charged.

Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCharging.java]: Thanks, all good now! I was looking at a stale commit 🤦

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/scope/HandleHederaOperations.java (hunk 6 above)
Comment (lukelee-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/scope/HandleHederaOperations.java:276]: it is called from here?  https://github.com/hiero-ledger/hiero-consensus-node/pull/20044/files#diff-d0ef7ec62e8802a84c068417e0c720f2761cf243b82d4cec3cf8f80682a51c4aR214

Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/scope/HandleHederaOperations.java:276]: Right, but then is `ProxyWorldUpdater.hasRefundGasFeeEvents` ever called other than in a test?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransaction.java (hunk 7 above)
Comment (lukelee-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransaction.java:129]: It turned out to be possible to get a negative number in `chargeWithRelayer` because the gas offered and the gas limit turned out to be a number much greater than the intrinsic gas.  Instead of adding a bunch of logic in that method, I felt it was simpler to force the logic into charging the relay account the intrinsic gas amount

Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransaction.java:129]: You mean this piece in the else block in `chargeWithRelayer`?
```
senderGasCost = transaction.offeredGasCost();
relayerGasCost = gasCost - transaction.offeredGasCost();
```

Actually this logic seems a bit weird and perhaps wrong 🤔 

Looking at that method:
if the sender offered a smaller gas price than `context.gasPrice()` (presumably the required price), then the sender pays the amount according to the `gasPrice` they've set, and the relayer pays the rest.

I think in the failure case we're handling in this PR this should be:
```
senderGasCost = intrinsicGas * txn.offeredGasPrice; // the gas we're charging * the price they're willing to pay per unit
relayerGasCost = gasCost - senderGasCost; // the remaining amount

// where gasCost is intrinsicGas * context.gasPrice(), i.e. the required amount
```
With the current logic we're still charging a gasLimit-based amount in this case.

So looks like we'll need to pass the gas amount to that method, rather than the cost to do this calculation.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/scope/HandleHederaOperations.java (hunk 6 above)
Comment (lukelee-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/scope/HandleHederaOperations.java:276]: It seemed to me that this could be useful in the future.  I'll get rid of it if you insist.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java (hunk 8 above)
Comment (lukelee-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java:257]: Yes, I agree that it would be better if all exceptions were handled the same way but it seems we want different behavior.  The other place where this exception can occur is if we run out storage slots which seems like a rare event and thus it seemed safe to lump in with the ops duration throttle exception.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransaction.java (hunk 7 above)
Comment (lukelee-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransaction.java:129]: I was hoping to reuse the same logic in `chargeWithRelayer`, hence my solution was to send in a 0 gas price to force the logic to a particular if clause in the case of an exception.

Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransaction.java:129]: This seems a bit...hacky 😄 Would there be any issues with updating `chargeWithRelayer` so that the logic is actually correct in all branches?

Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransaction.java:129]: And then revert the line where we set it to 0 to keep the original value.

I can easily imagine someone in the future trying to reference `offeredGasPrice` inside `maybeChargeFeesAndReturnOutcome` or some of its child calls - for whatever reason - and introducing a bug because they wouldn't expect it to be reset to 0.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/scope/HandleHederaOperations.java (hunk 6 above)
Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/scope/HandleHederaOperations.java:276]: I'm personally a proponent of the YAGNI principle 😄

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java (hunk 8 above)
Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ContextTransactionProcessor.java:257]: Fair enough 👍

**Comment by lukelee-sl** [comment]
closing and replacing with https://github.com/hiero-ledger/hiero-consensus-node/pull/20413
